### ISSUE TITLE
[Proposal] kubernetes cluster management using docker containers 

### ISSUE THREAD

**Issue by resouer** [kubernetes/kubernetes#13424]
### kubernetes cluster management using docker containers

cc @WIZARD-CXY  @dalanlan 

As we have already merged most of our internal code with 'docker-multinode' as planned, see: 
https://github.com/kubernetes/kubernetes/pull/11109
https://github.com/kubernetes/kubernetes/pull/12023
https://github.com/kubernetes/kubernetes/pull/10829

We can now update the current bare-metal cluster deployment with "kube-in-docker" methods, which will eliminate different OS distro, especially fix a batch of systemd related issues in Ubuntu 15.X (https://github.com/kubernetes/kubernetes/issues/8798)

So I propose it here to let people interested know. I already start the work from Ubuntu cluster.

/cc @roberthbailey @brendandburns @erictune 

[Update 2015-9-17]
- [x] Update existing docker-muitl scripts #13634 #13635
- [x] Manage ENVs in cluster
- [x] Implement SSH deployment & docs
- [ ] Review & refactoring #13901
- [ ] Merge with the current [docker-multi](https://github.com/kubernetes/kubernetes/tree/master/docs/getting-started-guides/docker-multinode) work

**Comment by resouer** [comment]
#dibs @dalanlan

**Comment by resouer** [comment]
#### Pros
1. No distro diversity anymore, this will close a lot of related area cluster issues
2. Much easier to maintain, one distro one manianer, that's wired. Eventually, have will only cloud provider and bare-metal provider
3. Much much easy to test & verify & tear down

#### Cons
1. Docker daemon will be used as a dependency, no plan to use rkt for now
2. Two daemons have to be running, but considering the simplification, docker-bootstrap have no process management for now
3. hyperkube will be used as binary, this is a little bit different for users who use k8s binaries directly. But we can also use k8s binaries eventually if everybody likes that.

**Comment by resouer** [comment]
#dibs @dalanlan @zju-sel

**Comment by gmarek** [comment]
cc @fgrzadkowski

**Comment by fgrzadkowski** [comment]
In general I really love the idea with running k8s in docker containers! However the problem with kube-in-docker setup is that does not support mounting any volumes. For the last week I've been trying to workaround this (proper fix on docker side is pending https://github.com/docker/docker/pull/15648#issuecomment-135035817). I managed to fix non-secret volumes (will send PR today), but there is still a problem with secrets (kubelet cannot write data to newly mounted tmpfs device for secrets).

Regarding having two deamons I think it should be possible to run only one deamon, but have to experiment more with this.

Related issue for running kubelet in container #4869.

**Comment by fgrzadkowski** [comment]
Another related issue #11984

**Comment by WIZARD-CXY** [comment]
I think it worth a try!

**Comment by dalanlan** [comment]
We can simply use flag `----privileged` to bypass the mounting issue in the short term, if that's what you concern most. @fgrzadkowski 
And here is what i'm getting via a docker-k8s.
```
$ kubectl exec busybox -- ls /var/run/secrets
kubernetes.io
```

**Comment by resouer** [comment]
Yes, actually we have been running `hyperkube kubelet` in privileged mode for a while, see: https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/docker-multinode/master.sh#L138. 

It is definitely a work round for now. And we must clarify all limits in docs until the mount issue is fixed in docker side.

We are also considered about two daemons, for now, we think a separate bootstrap daemon is good for segregation of duties, and can share burden from the main daemon

**Comment by fgrzadkowski** [comment]
I don't think that ```--priviliged``` changes anything. Container's rootfs is [rprivate](https://github.com/opencontainers/runc/blob/master/libcontainer/rootfs_linux.go#L411), by [default](https://github.com/docker/docker/blob/master/daemon/execdriver/driver_unix.go#L40). Please see https://github.com/docker/docker/pull/15648 which adds a flag to control this.

How did you setup your cluster? I checked it last week and it didn't work for me (@mwielgus reported the same in #11984)

**Comment by fgrzadkowski** [comment]
There is a RedHat distro version of docker which carries the SLAVE mount propagation patch (see [comment](https://github.com/kubernetes/kubernetes/issues/11984#issuecomment-127471040)). Maybe you are using it?

**Comment by fgrzadkowski** [comment]
@dalanlan btw - docker based cluster by default will not generate service account tokens, so I assume you must have been using non-standard setup.

**Comment by resouer** [comment]
@fgrzadkowski  I think the difference is in #11984 the author used single node guide from: http://kubernetes.io/v1.0/docs/getting-started-guides/docker.html , in which kubelet is started without privileged mode. 

But in original guides proposed by us & brendandburns, we used `--privileged` for all the kubelet container, see:
https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/docker-multinode.md

and

https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/docker.md

Could you please test with that doc & scripts if possible?

**Comment by fgrzadkowski** [comment]
I checked with docs from HEAD and ```--privileged``` flag. It doesn't work. I linked some issues/prs (see above) which explain why it doesn't work. tl;dr: container's rootfs is private by default, so mounts created by kubelet in a container does not propagate to the host and thus are not copied to the newly created namespace for a new container. There is an experimental code in ```nsenter_mounter.go``` that fixes some problems with mounting volumes (I'm just sending a PR to fix some minor issues there), but it doesn't solve the problem with secrets, because we need to write data to this newly created mount.

**Comment by resouer** [comment]
@fgrzadkowski I think I catch up some of your points now. Could you please give us a example, like a pod.yml for us to reproduce the problem? As it seems our env "works" some how ... I'd like to dig it out.

OTOH, I think we can also mount the entire host filesystem inside kubelet, what do you think?

**Comment by dalanlan** [comment]
Fair point. I'd done so many tricks that i could not remember clearly, i'll give you that. @fgrzadkowski 
I‘ll try from HEAD and see how it goes later (maybe after #13446 is merged, i guess).

**Comment by fgrzadkowski** [comment]
@resouer How to reproduce:

```
filipg@vm-ubuntu:~$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
filipg@vm-ubuntu:~$ docker run --net=host -d gcr.io/google_containers/etcd:2.0.12 /usr/local/bin/etcd --addr=127.0.0.1:4001 --bind-addr=0.0.0.0:4001 --data-dir=/var/etcd/data
5787bfe15431844f5d2feb59d4c29a7cde461b5fc519dc04560c95c818caf3f0
filipg@vm-ubuntu:~$ docker run --net=host --privileged -d -v /sys:/sys:ro -v /var/run/docker.sock:/var/run/docker.sock  gcr.io/google_containers/hyperkube:v1.0.1 /hyperkube kubelet --api-servers=http://localhost:8080 --v=2 --address=0.0.0.0 --enable-server --hostname-override=127.0.0.1 --config=/etc/kubernetes/manifests
10223ec1a58d71cf3c5752eca01851c72f4774d65d22f7bd70a246124fcdf4bf
filipg@vm-ubuntu:~$ docker run -d --net=host --privileged gcr.io/google_containers/hyperkube:v1.0.1 /hyperkube proxy --master=http://127.0.0.1:8080 --v=2
040d5b059e4778b65d0fe0f1c1f9c8c57e75047e2447bc44c7780948348cd1a1
filipg@vm-ubuntu:~$ docker ps
CONTAINER ID        IMAGE                                       COMMAND                  CREATED              STATUS              PORTS               NAMES
ae291de2bda1        gcr.io/google_containers/hyperkube:v1.0.1   "/hyperkube scheduler"   57 seconds ago       Up 56 seconds                           k8s_scheduler.2744e742_k8s-master-127.0.0.1_default_f3ccbffbd75e3c5d2fb4ba69c8856c4a_9dcd9a01
84655dfa8cde        gcr.io/google_containers/hyperkube:v1.0.1   "/hyperkube apiserver"   57 seconds ago       Up 56 seconds                           k8s_apiserver.cfb70250_k8s-master-127.0.0.1_default_f3ccbffbd75e3c5d2fb4ba69c8856c4a_f03fb3dd
13cc50218cb6        gcr.io/google_containers/hyperkube:v1.0.1   "/hyperkube controlle"   57 seconds ago       Up 56 seconds                           k8s_controller-manager.1598ee5c_k8s-master-127.0.0.1_default_f3ccbffbd75e3c5d2fb4ba69c8856c4a_0a90b76c
040d5b059e47        gcr.io/google_containers/hyperkube:v1.0.1   "/hyperkube proxy --m"   About a minute ago   Up About a minute                       focused_goodall
9f6b3b820c44        gcr.io/google_containers/pause:0.8.0        "/pause"                 About a minute ago   Up About a minute                       k8s_POD.e4cc795_k8s-master-127.0.0.1_default_f3ccbffbd75e3c5d2fb4ba69c8856c4a_b2b0388d
10223ec1a58d        gcr.io/google_containers/hyperkube:v1.0.1   "/hyperkube kubelet -"   About a minute ago   Up About a minute                       cranky_leakey
5787bfe15431        gcr.io/google_containers/etcd:2.0.12        "/usr/local/bin/etcd "   About a minute ago   Up About a minute                       stoic_mestorf
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 get pods --all-namespaces
NAMESPACE   NAME                   READY     STATUS    RESTARTS   AGE
default     k8s-master-127.0.0.1   3/3       Running   0          1m
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 get pods --all-namespaces
NAMESPACE   NAME                   READY     STATUS    RESTARTS   AGE
default     k8s-master-127.0.0.1   3/3       Running   0          1m
filipg@vm-ubuntu:~$ cat pod.json 
{
  "apiVersion": "v1",
  "kind": "Pod",
  "metadata": {
    "name": "mypod"
  },
  "spec": {
    "containers": [{
      "name": "mypod",
      "image": "nginx",
      "volumeMounts": [{
        "name": "foo",
        "mountPath": "/etc/foo",
        "readOnly": true
      }]
    }],
    "volumes": [{
      "name": "foo",
      "secret": {
        "secretName": "mysecret"
      }
    }]
  }
}
filipg@vm-ubuntu:~$ cat secret.json 
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
data:
  password: dmFsdWUtMg0K
  username: dmFsdWUtMQ0K
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 create -f secret.json 
secrets/mysecret
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 create -f pod.json 
pods/mypod
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 get pods --all-namespaces
NAMESPACE   NAME                   READY     STATUS    RESTARTS   AGE
default     k8s-master-127.0.0.1   3/3       Running   0          2m
default     mypod                  0/1       Running   0          15s
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 get secrets --all-namespaces
NAMESPACE   NAME       TYPE      DATA
default     mysecret   Opaque    2
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 exec mypod -- ls /etc/foo
filipg@vm-ubuntu:~$ kubectl -s http://localhost:8080 exec mypod -- ls /etc | grep foo
foo
```

This is running on GCE VM with Ubuntu 15.04 (but I also checked with Ubuntu 14.10). I have a similar example with GCE PD.

Regarding mounting the entire host filesystem inside kubelet, this will also not help. New mounts will not be propagated anyway.

BTW - I've just realized that listing directory (```ls /var/run/secrets``` in your example) will indeed work, because the directory will be created and indeed flag ```--privileged``` is needed for that. It's just that the directory will be empty, because nothing will be mounted there.

**Comment by resouer** [comment]
@fgrzadkowski  thanks for all your work. me & dalanlan also did a lot of experiments and confirm your issues: the dir is empty. sorry for our carelessness.

hostPath is not influenced as it is send to dockerd to process (correct me if I'm wrong), but other volume types will suffer mount problem.    

Now I can followed up all your points when collecting [this blog](https://huaminchen.wordpress.com/2015/06/09/share-hosts-mount-namespace-with-docker-containers/) and https://github.com/docker/docker/pull/15648 all together.

docker 1.9 is too far away honestly ...

**Comment by roberthbailey** [comment]
@dchen1107

**Comment by kelseyhightower** [comment]
I've also gone down this path many times, but all the workarounds seemed like overkill. After some time I've come to the conclusion that the Kubelet "owns" the machine. When I stop to consider all the things the Kubelet is responsible for, then it starts to make less sense to have docker or any container run time as a dependency for the Kubelet.

With support for network plugins, it may very well come a day where the Kubelet only uses Docker, or maybe something line runc, to spin up containers using the volumes and networks created by the Kubelet.

**Comment by fgrzadkowski** [comment]
@kelseyhightower I believe that main advantage of using docker for kubelet is that it's more portable - we have all the files, configuration etc bound together.

**Comment by resouer** [comment]
@kelseyhightower yes, portable, on the other hand, eliminate OS distro. For now, we have to not only take care of upstart, systemd, but also a huge bunch of scripts to adapt to different OSes. I have already done a POC, which is only half LOC of current ubuntu-cluster scripts, but take care of all distros. 

A compromise is we can run kubelet in baremetal (have to take care both upstart & systemd ) but master in docker (I used to do this months ago), and wait for issues fixed.

**Comment by kelseyhightower** [comment]
@fgrzadkowski @resouer If the kubelet can run from a docker container that would be awesome, but it seems to require a hard dependance on Docker and ongoing support from the upstream docker project.

**Comment by fgrzadkowski** [comment]
@resouer Somehow I've ignored the fact that you are referring here only to bare metal setup. I'd like us to go into the direction where docker based setup is the default for most providers (bare metal, GCE, AWS and all OS distributions). I think there is not reason to limit this to bare metal.

**Comment by resouer** [comment]
@fgrzadkowski  can't agree more!

**Comment by luxas** [comment]
@fgrzadkowski Is any progress made on this topic?
I'm very interested in running everything in docker.
That is mostly because I run things on ARM, see my [project](https://github.com/luxas/kubernetesonarm)

Things are working, but e.g. the default serviceAccount isn't automatically made.

**Comment by fgrzadkowski** [comment]
We are now discussing (with @roberthbailey and @mikedanese and others) how we want to refactor kube-up scripts to make them more pluggable. Once we have this we will combine it with docker based setup. This should make it quite easy to run it wherever by just overriding certain parts of the scripts.

This however is going quite slow as our main focus now is shipping 1.1.

**Comment by resouer** [comment]
@luxas you can follow the progress of #13901, I guess ...

**Comment by luxas** [comment]
Thanks for the link. Do anyone know why serviceAccounts doesn't work by default when apiserver is run in a pod?

**Comment by dalanlan** [comment]
Some param of apiserver should be properly set to enable SA, `--service-account-key-file` or `--tls-private-key-file` for instance. 
Yet you'd better turn to stackoverflow for this.

**Comment by resouer** [comment]
@luxas api-server in pod load a [master-multi.json](https://github.com/kubernetes/kubernetes/blob/master/cluster/images/hyperkube/master-multi.json) from hyperkube image, which, is not service-account configured by default. 

That's why in #13901 we proposed a way to load user provided master-multi.json, so you can configure that freely. See: [this doc](https://github.com/kubernetes/kubernetes/pull/13901/files#diff-887eb6de14d1a695be41363b2ece2c34R180)

**Comment by WIZARD-CXY** [comment]
hmm，good point

May the force be with you

**Comment by fgrzadkowski** [comment]
@luxas @resouer Actually it's not about the configuration, but it's because secrets doesn't work in the released version of hyperkube. I have fixed at HEAD, but didn't have time to release it.

**Comment by luxas** [comment]
@resouer Yes, I have [master-multi.json](https://github.com/luxas/kubernetes-on-arm/blob/master/images/kubernetesonarm/hyperkube/master-multi.json) in my custom built ARM hyperkube.
And everything is working, except for secrets in  `seviceAccounts`. 

@fgrzadkowski Could you point our the commit where you fixed it? Is it coming in the v1.1 release?

Thanks for your time.

**Comment by fgrzadkowski** [comment]
@luxas Please see https://github.com/kubernetes/kubernetes/issues/13557. Apart from the PR listed there you'll also need to use ```--containerized``` flag (it's explained in the issue)

**Comment by dalanlan** [comment]
I built one from HEAD on Oct 16th, yet i got following
```
$ kubectl get serviceAccounts --all-namespaces
NAMESPACE   NAME      SECRETS
default     default   0
$ kubectl get serviceAccounts/default -o yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: 2015-10-16T06:34:57Z
  name: default
  namespace: default
  resourceVersion: "19"
  selfLink: /api/v1/namespaces/default/serviceaccounts/default
  uid: 050c86d9-73d0-11e5-a924-000c2994ac1c
$ kubectl get secrets --all-namespaces
NAMESPACE   NAME      TYPE      DATA

```
No secret ref by SA `default`. Is this expected then?
I was completely fine with that earlier since i was thinking secrets won't be generated w/o proper params for api-server.

**Comment by bgrant0607** [comment]
I think @pmorie did some work on making Kubelet run in a Docker container a while back.

**Comment by bgrant0607** [comment]
#4869

**Comment by luxas** [comment]
@fgrzadkowski 
I downloaded `v1.2.0-alpha.3` binaries, copied `hyperkube` to cluster/images/hyperkube and built that image with the right tag.
Then I copied the commands from `docs/getting-started-guides/docker.md` and some secs later the master was up.

But, `kubectl get secrets` doesn't return anything. Exactly the same result as @dalanlan
I'm quite sure that your commit was included in that version.
Some thoughts? Or does one now have to configure the apiserver correctly?

**Comment by pmorie** [comment]
@kubernetes/rh-cluster-infra

**Comment by dalanlan** [comment]
My guess is that we do need `--service-account-key-file` or `--tls-private-key-file`. (didn't follow this thread for a while, though)

**Comment by luxas** [comment]
@dalanlan An interesting thing is that `docker exec <apiserver> ls /var/run/kubernetes` returns two files: 
 - apiserver.key
 - apiserver.crt

apiserver help (v1.2.0-alpha.3):
```
      --tls-cert-file="": File containing x509 Certificate for HTTPS.  (CA cert, if any, concatenated after server cert). If HTTPS serving is enabled, and --tls-cert-file and --tls-private-key-file are not provided, a self-signed certificate and key are generated for the public address and saved to /var/run/kubernetes.
      --tls-private-key-file="": File containing x509 private key matching --tls-cert-file.
      --service-account-key-file="": File containing PEM-encoded x509 RSA private or public key, used to verify ServiceAccount tokens. If unspecified, --tls-private-key-file is used.
```
`docker logs <apiserver>`
```
server.go:620] Serving securely on 0.0.0.0:6443
server.go:632] Using self-signed cert (/var/run/kubernetes/apiserver.crt, /var/run/kubernetes/apiserver.key)
```
WDYT?
/cc @fgrzadkowski

**Comment by luxas** [comment]
Now I compared `hack/local-up-cluster.sh` to `docker.md` above

When running `hack/local-up-cluster.sh`, the tmpfs is created correctly:
```
mount (on host)
tmpfs on /var/lib/kubelet/pods/6160ea3c-8538-11e5-855f-f0761c62f136/volumes/kubernetes.io~secret/default-token-1cmtw type tmpfs (rw,relatime)

kubectl exec busybox -- mount
tmpfs on /var/run/secrets/kubernetes.io/serviceaccount type tmpfs (ro,relatime)
```

But, no such mounts when running in docker. 
It seems like dockerized kubelet can't mount `tmpfs` on host's `-v /var/lib/kubelet:/var/lib/kubelet:rw`.
Another thing is that dockerized kubelet doesn't either make the `kubernetes.io~secret` dir. Maybe because the apiserver flag is missing
```
# when running on host
ls /var/lib/kubelet/pods/6160ea3c-8538-11e5-855f-f0761c62f136/volumes
kubernetes.io~secret

# when running in docker
ls /var/lib/kubelet/pods/81aa896a-853c-11e5-a8f8-f0761c62f136/volumes
(no dir)
```

**Comment by guybrush** [comment]
@luxas i think these are related: #13791 #16055

as far as i understand you might be able to get it running by using a hyperkube image built from current master (but i did not try it myself yet)

**Comment by dalanlan** [comment]
It should be since #13791 was merged (Sep 18th), yet i built one from HEAD on Oct 16th, only to find it did not.
I find it necessary to file another issue to track this. Stay tuned.

**Comment by dalanlan** [comment]
#16965 manage to track this issue.

**Comment by dalanlan** [comment]
And it has nothing to do with the param configuration. Sorry for my misleading, if there is any.

**Comment by luxas** [comment]
@guybrush As I said [above](https://github.com/kubernetes/kubernetes/issues/13424#issuecomment-154548020), I used `v1.2.0-alpha.3`, where #13791 is merged. That patch doesn't seem to fix it 

We'll track this issue in @dalanlan's #16965 from now on I think

**Comment by guybrush** [comment]
ah sorry, i totally missed that comment :D

**Comment by luxas** [comment]
Yeah, that happens :)

**Comment by mingfang** [comment]
I have been running the Master and Nodes inside Docker for many months now.
Instructions can be found here https://github.com/mingfang/diy-kubernetes
Note it has the same volume mounting limitations.

**Comment by luxas** [comment]
@mingfang Yeah, me too. On Raspberry Pi 2s. My project is here: https://github.com/luxas/kubernetes-on-arm
But it would be great to be able to use `secrets`, `volumes` and `serviceAccounts` too.
Because of that, we are trying to solve these problems: #16965. #13791 already fixed `volumes`

**Comment by pmorie** [comment]
@mingfang You need to use `--containerized` when running kubelet in container -- see `local-up-cluster.sh`

**Comment by mingfang** [comment]
@pmorie I tried ```--containerized``` but since secrets still does not work using it in v1.1.1-beta.1, I'm not sure what other benefits would it give me.
Also when using ```--containerized```, running this command ```nsenter --mount=/rootfs/proc/1/ns/mnt -- /bin/findmnt --target /var/lib/kubelet ``` always returns nothing.
Is there a trick to get ```--containerized``` to work?

**Comment by dalanlan** [comment]
@mingfang  Pull from HEAD and give it a try

**Comment by pmorie** [comment]
@mingfang Did you bind-mount `/proc` on the host to `/rootfs/proc` in the kubelet container?

**Comment by mingfang** [comment]
@pmorie I think so.  This is the relevant parts of run script,
```
  docker run \
         --net=host \
         --pid=host \
         --privileged=true \
         -v /sys:/sys:ro \
         -v /var/run:/var/run:rw \
         -v /:/rootfs:ro \
         -v /dev:/dev \
         -v /var/run/docker.sock:/run/docker.sock \
         -v /var/lib/docker/:/var/lib/docker:rw \
         -v /var/lib/kubelet/:/var/lib/kubelet:rw \
         -v /var/run/openvswitch/db.sock:/var/run/openvswitch/db.sock \
         -d kubernetes-node
```

**Comment by pmorie** [comment]
@mingfang Could you run the secrets e2e against the containerized kubelet and post the kubelet log from that run?  That will help us debug.

**Comment by mingfang** [comment]
@pmorie Sorry I don't know how to run these tests.
Here is more info that I hope is helpful.

Host
```
uname -a
Linux vm3 3.13.0-63-generic #103-Ubuntu SMP Fri Aug 14 21:42:59 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```
findmnt from host
```
# findmnt
TARGET                                                                                               SOURCE                                                              FSTYPE     OPTIONS
/                                                                                                    /dev/disk/by-uuid/c9efd6cc-5c65-404e-937f-9447fa63038d              ext4       rw,relatime,data=or
├─/sys                                                                                               sysfs                                                               sysfs      rw,nosuid,nodev,noe
│ ├─/sys/fs/cgroup                                                                                                                                                       tmpfs      rw,relatime,size=4k
│ │ ├─/sys/fs/cgroup/cpuset                                                                          cgroup                                                              cgroup     rw,relatime,cpuset
│ │ ├─/sys/fs/cgroup/cpu                                                                             cgroup                                                              cgroup     rw,relatime,cpu
│ │ ├─/sys/fs/cgroup/cpuacct                                                                         cgroup                                                              cgroup     rw,relatime,cpuacct
│ │ ├─/sys/fs/cgroup/memory                                                                          cgroup                                                              cgroup     rw,relatime,memory
│ │ ├─/sys/fs/cgroup/devices                                                                         cgroup                                                              cgroup     rw,relatime,devices
│ │ ├─/sys/fs/cgroup/freezer                                                                         cgroup                                                              cgroup     rw,relatime,freezer
│ │ ├─/sys/fs/cgroup/blkio                                                                           cgroup                                                              cgroup     rw,relatime,blkio
│ │ ├─/sys/fs/cgroup/perf_event                                                                      cgroup                                                              cgroup     rw,relatime,perf_ev
│ │ ├─/sys/fs/cgroup/hugetlb                                                                         cgroup                                                              cgroup     rw,relatime,hugetlb
│ │ └─/sys/fs/cgroup/systemd                                                                         systemd                                                             cgroup     rw,nosuid,nodev,noe
│ ├─/sys/fs/fuse/connections                                                                                                                                             fusectl    rw,relatime
│ ├─/sys/kernel/debug                                                                                                                                                    debugfs    rw,relatime
│ ├─/sys/kernel/security                                                                                                                                                 securityfs rw,relatime
│ └─/sys/fs/pstore                                                                                                                                                       pstore     rw,relatime
├─/proc                                                                                              proc                                                                proc       rw,nosuid,nodev,noe
├─/dev                                                                                               udev                                                                devtmpfs   rw,relatime,size=20
│ └─/dev/pts                                                                                         devpts                                                              devpts     rw,nosuid,noexec,re
├─/run                                                                                               tmpfs                                                               tmpfs      rw,nosuid,noexec,re
│ ├─/run/lock                                                                                                                                                            tmpfs      rw,nosuid,nodev,noe
│ ├─/run/shm                                                                                                                                                             tmpfs      rw,nosuid,nodev,rel
│ ├─/run/user                                                                                                                                                            tmpfs      rw,nosuid,nodev,noe
│ └─/run/docker/netns/default                                                                        proc                                                                proc       rw,nosuid,nodev,noe
├─/var/lib/docker/aufs                                                                               /dev/disk/by-uuid/c9efd6cc-5c65-404e-937f-9447fa63038d[/var/lib/docker/aufs]
                                                                                                                                                                         ext4       rw,relatime,data=or
│ └─/var/lib/docker/aufs/mnt/91d838ebeee66d10b359505f8dbcd07674743f5b32ddc611a96c49de92d43938                                                                            aufs       rw,relatime,si=2473
├─/var/lib/docker/containers/91d838ebeee66d10b359505f8dbcd07674743f5b32ddc611a96c49de92d43938/shm    shm                                                                 tmpfs      rw,nosuid,nodev,noe
└─/var/lib/docker/containers/91d838ebeee66d10b359505f8dbcd07674743f5b32ddc611a96c49de92d43938/mqueue mqueue                                                              mqueue     rw,nosuid,nodev,noe
```
Inside kubelet
```
kubelet --version=true
Kubernetes v1.2.0-alpha.3
```
Test secret(relevant parts only)
```
kubectl get secrets -o yaml
items:
- apiVersion: v1
  data:
    id-rsa: dmFsdWUtMg0KDQo=
    id-rsa.pub: dmFsdWUtMQ0K
  kind: Secret
  metadata:
    name: ssh-key-secret
```
RC using test secret
```
kubectl get rc redis -o yaml
kind: ReplicationController
spec:
  template:
    spec:
      containers:
      - image: redis:latest
        volumeMounts:
        - mountPath: /var/log
          name: log
        - mountPath: /etc/ssh-keys
          name: ssh-keys
          readOnly: true
      volumes:
      - emptyDir: {}
        name: log
      - name: ssh-keys
        secret:
          secretName: ssh-key-secret
```
This is the kind of errors I'm seeing in /var/log/kubelet
```
2015-11-08_06:53:38.06371 I1108 06:53:38.063678    2202 volumes.go:109] Used volume plugin "kubernetes.io/secret" for ssh-keys
2015-11-08_06:53:38.06379 I1108 06:53:38.063772    2202 nsenter_mount.go:174] findmnt command: nsenter [--mount=/rootfs/proc/1/ns/mnt -- /bin/findmnt -o target --noheadings --target /var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys]
2015-11-08_06:53:38.06642 I1108 06:53:38.066363    2202 secret.go:131] Setting up volume ssh-keys for pod 12236f49-85e1-11e5-a0c3-7824afc0a170 at /var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys
2015-11-08_06:53:38.06647 I1108 06:53:38.066464    2202 volumes.go:109] Used volume plugin "kubernetes.io/empty-dir" for 
2015-11-08_06:53:38.06651 I1108 06:53:38.066503    2202 nsenter_mount.go:174] findmnt command: nsenter [--mount=/rootfs/proc/1/ns/mnt -- /bin/findmnt -o target --noheadings --target /var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys]
2015-11-08_06:53:38.06843 I1108 06:53:38.068414    2202 empty_dir_linux.go:39] Determining mount medium of /var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys
2015-11-08_06:53:38.06846 I1108 06:53:38.068455    2202 nsenter_mount.go:174] findmnt command: nsenter [--mount=/rootfs/proc/1/ns/mnt -- /bin/findmnt -o target --noheadings --target /var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys]
2015-11-08_06:53:38.07020 E1108 06:53:38.070182    2202 kubelet.go:1480] Unable to mount volumes for pod "redis-muzm8_default": IsLikelyNotMountPoint("/var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys"): file does not exist; skipping pod
2015-11-08_06:53:38.07028 I1108 06:53:38.070265    2202 kubelet.go:2873] Generating status for "redis-muzm8_default"
2015-11-08_06:53:38.07396 I1108 06:53:38.072864    2202 server.go:761] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"redis-muzm8", UID:"12236f49-85e1-11e5-a0c3-7824afc0a170", APIVersion:"v1", ResourceVersion:"2582041", FieldPath:""}): reason: 'FailedMount' Unable to mount volumes for pod "redis-muzm8_default": IsLikelyNotMountPoint("/var/lib/kubelet/pods/12236f49-85e1-11e5-a0c3-7824afc0a170/volumes/kubernetes.io~secret/ssh-keys"): file does not exist
```

Please let me know if I can provide anymore information.

**Comment by dalanlan** [comment]
@mingfang  Ref https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#end-to-end-tests on how to run e2e tests.

**Comment by dalanlan** [comment]
It could be a little messy, yet i have reason to believe that containerized kubelet works with secrets now (at least from HEAD, not sure about specific release, cherry-pick thing).

Secrets would not be generated automatically until you enable flags for controller-manager and api-server (`--admission-control`). 

Feel free to ref @luxas's work specified [#16965(comment)](https://github.com/kubernetes/kubernetes/issues/16965#issuecomment-154740451), or our work on bare-metal ubuntu (use easyrsa tool to generate a key file instead of a self-signed one). 

Hope that helps.

**Comment by luxas** [comment]
@dalanlan @mingfang `git describe --contains 7fe34f205020874885a1063f471ede4a5bee3176` returns `v1.2.0-alpha.1~204^2`, so at least in `v1.2.0` #13791 is included (and HEAD, of course)

TL;DR https://github.com/kubernetes/kubernetes/issues/16965#issuecomment-154740451
 - apiserver `--admission-control=ServiceAccount` is required for `/var/run/secrets` in containers.
 - #13791 needs to be merged for volumes to work
 -  `controller-manager` needs the same `.key` and `.crt` files as `apiserver`. Apiserver self-signs by default.
 - this can be fixed by adding a volume to `master-{,multi}.json`, so the master components share the certs.

**Comment by mingfang** [comment]
How important is setting ```--resource-container=""```? ref https://github.com/kubernetes/kubernetes/issues/12386

**Comment by mingfang** [comment]
In the kubelet logs I'm seeing these messages 
```docker.go:369] Docker Container: /sharp_hodgkin is not managed by kubelet.``
In this case, ```/sharp_hodgkin``` is the name of the kubelet container itself.
```
docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
7e284813ba21        kubernetes-node     "/bin/sh -c 'export >"   About an hour ago   Up About an hour                        sharp_hodgkin
```
There are no apparent adverse effects other than populating the logs though.

**Comment by pmorie** [comment]
@mingfang That shouldn't be a problem.

**Comment by luxas** [comment]
@mingfang Do you have `--v=3`? 
Decrementing the log level makes you get rid of the msgs.
But, as @pmorie say, no problem.

**Comment by mingfang** [comment]
@luxas @pmorie Yes my log level was ```--v=5``` because I was trying to debug the volume mounting issues.  I can confirm that there are no more printouts with ```--v=2```.  Thanks for  confirming that this is a non-issue.

**Comment by errordeveloper** [comment]
@kubernetes/sig-cluster-lifecycle for the start, this had been inactive for quite a while. And, in the light of recent activity on https://github.com/kubernetes/features/issues/11, should could close this (?), as the use-case here is about simplifying deployment, which is what we are solving now without having to run kubelet in Docker.

**Comment by mikedanese** [comment]
ya, docker multinode exists, so this can be closed.
